fuzzers/libafl-fuzz/src/fuzzer.rs

```diff
@@ -0,0 +1,395 @@
+    schedulers::{
+        powersched::PowerSchedule, IndexesLenTimeMinimizerScheduler, StdWeightedScheduler,
+    },
+    stages::{
+        mutational::MultiMutationalStage, CalibrationStage, ColorizationStage, IfStage,
+        StagesTuple, StdPowerMutationalStage,
+    },
+    state::{
+        HasCorpus, HasCurrentTestcase, HasExecutions, HasLastReportTime, HasStartTime, StdState,
+        UsesState,
+    },
+    Error, Fuzzer, HasFeedback, HasMetadata,
+};
+use libafl_bolts::{
+    current_nanos, current_time,
+    fs::get_unique_std_input_file,
+    ownedref::OwnedRefMut,
+    rands::StdRand,
+    shmem::{ShMem, ShMemProvider, StdShMemProvider},
+    tuples::{tuple_list, Handled, Merge},
+    AsSliceMut,
+};
+use libafl_targets::{cmps::AFLppCmpLogMap, AFLppCmpLogObserver, AFLppCmplogTracingStage};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    afl_stats::AflStatsStage,
+    corpus::{set_corpus_filepath, set_solution_filepath},
+    env_parser::AFL_DEFAULT_MAP_SIZE,
+    feedback::{filepath::CustomFilepathToTestcaseFeedback, seed::SeedFeedback},
+    Opt, AFL_DEFAULT_INPUT_LEN_MAX, AFL_DEFAULT_INPUT_LEN_MIN, SHMEM_ENV_VAR,
+};
+
+pub fn run_client<EMH, SP>(
+    state: Option<LibaflFuzzState>,
+    mut restarting_mgr: CentralizedEventManager<
+        LlmpRestartingEventManager<(), LibaflFuzzState, SP>,
+        EMH,
+        LibaflFuzzState,
+        SP,
+    >,
+    fuzzer_dir: &PathBuf,
+    opt: &Opt,
+) -> Result<(), Error>
+where
+    EMH: EventManagerHooksTuple<LibaflFuzzState> + Copy + Clone,
+    SP: ShMemProvider,
+{
+    // Create the shared memory map for comms with the forkserver
+    let mut shmem_provider = StdShMemProvider::new().unwrap();
+    let mut shmem = shmem_provider
+        .new_shmem(opt.map_size.unwrap_or(AFL_DEFAULT_MAP_SIZE))
+        .unwrap();
+    shmem.write_to_env(SHMEM_ENV_VAR).unwrap();
+    let shmem_buf = shmem.as_slice_mut();
+
+    // Create an observation channel to keep track of edges hit.
+    let edges_observer = unsafe {
+        HitcountsMapObserver::new(StdMapObserver::new("edges", shmem_buf)).track_indices()
+    };
+
+    // Create a MapFeedback for coverage guided fuzzin'
+    let map_feedback = MaxMapFeedback::new(&edges_observer);
+
+    // Create the CalibrationStage; used to measure the stability of an input.
+    let calibration = CalibrationStage::new(&map_feedback);
+
+    // Create an observation channel to keep track of the execution time.
+    let time_observer = TimeObserver::new("time");
+
+    /*
+     * Feedback to decide if the Input is "corpus worthy"
+     * We only check if it gives new coverage.
+     * The `TimeFeedback` is used to annotate the testcase with it's exec time.
+     * The `CustomFilepathToTestcaseFeedback is used to adhere to AFL++'s corpus format.
+     * The `Seedfeedback` is used during seed loading to adhere to AFL++'s handling of seeds
+     */
+    let mut feedback = SeedFeedback::new(
+        feedback_or!(
```

Comment:
The `feedback_or` macro seems to take multiple parguments, so no need to nest them I think.